assets/app/scripts/services/data.js

```diff
@@ -199,15 +125,269 @@ angular.module('openshiftConsole')
   };
 
 
-  DataService.prototype._stopListeningForUpdates = function(type) {
-    if (this._openConnections[type]) {
-     this._openConnections[type].close();
-     // TODO can we use delete here instead, or will that screw up the onclose event
-     this._openConnections[type] = null;
+  function DataService() {
+    this._listCallbacksMap = {};
+    this._watchCallbacksMap = {};
+    this._watchOperationMap = {};
+    this._listOperationMap = {};
+    this._resourceVersionMap = {};
+    this._dataMap = {};
+    this._watchWebsocketsMap = {};
+    this._watchPollTimeoutsMap = {};
+  }
+
+// type:      API type (e.g. "pods")
+// context:   API context (e.g. {project: "..."})
+// callback:  function to be called with the initial list of the requested type,
+//            and when updates are received, parameters passed to the callback:
+//            list:   current list of objects (context-qualified)
+//                    represented as a list (to preserve server-order)
+//                    helper method for returning a map indexed by attribute (e.g.
+//                    list.by('metadata.uid'))
+// opts:      options (currently none, placeholder)
+  DataService.prototype.list = function(type, context, callback, opts) {
+    this._listCallbacks(type, context).add(callback);
+
+    if (this._watch(type, context) && this._resourceVersion(type, context)) {
+      this._listCallbacks(type, context).fire(this._data(type, context),null,null);
```

Comment:
maybe a comment about what this means... `// A watch operation is running, and we've received the initial set of data for this type`